doc/threat_model/README.md

```diff
@@ -0,0 +1,157 @@
+Tock Threat Model
+=================
+
+## Overview
+
+Tock provides hardware-based isolation between applications as well as
+language-based isolation between untrusted kernel capsules.
+
+Tock supports a variety of hardware, including boards defined in the Tock
+repository and boards defined "out of tree" in a separate repository.
+Additionally, Tock's installation model may vary between different use cases
+even when those use cases are based on the same hardware. As a result of Tock's
+flexibility, the mechanisms it uses to provide isolation — and the strength of
+that isolation — vary from deployment to deployment.
+
+This threat model describes the isolation provided by Tock as well as the trust
+model that Tock uses to implement that isolation. Users of Tock, which include
+board integrators and application developers, should use this threat model to
+understand what isolation Tock provides to them (and what isolation it may not
+provide). Tock developers should use this threat model as a guide for how to
+provide Tock's isolation guarantees.
+
+## Definitions
+
+These definitions are shared between the documents in this directory.
+
+**Application data** includes an application's image in non-volatile storage,
+its memory footprint in RAM, and any data that conceptually belongs to the
+application that is held by the kernel or other applications. For example, if an
+application requests samples from an ADC then those samples are considered the
+application's data, even when they are stored in a location in RAM only readable
+by the kernel.
```

Comment:
Hmm, that's a question that seems to have a simple answer but which is nontrivial to word precisely. My intuition is that if 2 applications request a reading from the same ADC concurrently then the ADC virtualization is take a single sample and distribute it to both applications. However, the text I currently have in `Virtualization.md` seems to ban that implementation.

Perhaps instead of having a "Shared Buses" section, we should have a "Shared Inputs" section. Then an ADC and an un-muxed UART would both count as "shared inputs" and would be subject to the same rules about isolation. That would allow capsules to copy a single reading to multiple apps.

Reply:
> My intuition is that if 2 applications request a reading from the same ADC concurrently then the ADC virtualization is take a single sample and distribute it to both applications. However, the text I currently have in Virtualization.md seems to ban that implementation.

ADC is maybe a bit more subtle to think about, but _definitely_ if two applications request a temperature reading shortly one after the other, the kernel driver should be able to only take one actual reading.

Reply:
In general, this is not inconsistent _so far_ (later on in the document it gets a bit trickier). Data has a unique "owner" _by default_, but data could belong to several applications at once. In particular, both memory shared between processes and memory that resides in the kernel (a sensor reading) can be "Application data" for multiple processes.

Reply:
Based on the core WG call last Friday, I changed this to clarify that ADC drivers (and similar drivers) may make readings and share them with multiple clients.

I think there will always be a judgement call here, but one distinction I see that differentiates it from the UART is that the ADC driver receives requests from its clients to take samples whereas the UART driver receives interrupts from the driver (i.e. push vs. pull model). I'm not sure if that difference will always line up with our isolation desires.